drivers/spi/spi_sy1xx.c

```diff
@@ -0,0 +1,506 @@
+	buf[count++] = SPI_CMD_READ_DATA1();
+	buf[count++] = SPI_CMD_READ_DATA0(0, 0);
+
+	/* end of transmission */
+	buf[count++] = SPI_CMD_EOT3(0);
+	buf[count++] = SPI_CMD_EOT2();
+	buf[count++] = SPI_CMD_EOT1();
+	buf[count++] = SPI_CMD_EOT0();
+
+	SY1XX_UDMA_START_RX(config->base, (uint32_t)data->read, request->data_len,
+			    SY1XX_UDMA_RX_DATA_ADDR_INC_SIZE_32);
+	SY1XX_UDMA_START_TX(config->base, (uint32_t)buf, count, 0);
+
+	SY1XX_UDMA_WAIT_FOR_FINISHED_TX(config->base);
+	SY1XX_UDMA_WAIT_FOR_FINISHED_RX(config->base);
+
+	int32_t xfer_count = request->data_len - SY1XX_UDMA_GET_REMAINING_RX(config->base);
+
+	if ((xfer_count > 0) && (xfer_count <= request->data_len)) {
+		memcpy(request->data, &data->read[0], xfer_count);
+		return xfer_count;
+	}
+
+	return 0;
+}
+
+/**
+ * the problem with transfer function is, that it comes in
+ * multiple chunks. where we actually want to use DMA to transfer all address + data
+ * at once.
+ *
+ * transfers could be,
+ * - read (data)
+ * - write (data)
+ * - write (addr), read (data)
+ * - write (addr), write (data)
+ *
+ * DMA can distinguish between address and data writing, so we
+ * have to de-puzzle the addr and data from tx_bufs and rx_bufs.
+ *
+ * @return
+ *  - < 0: Error
+ *  -   0: OK
+ */
+static int sy1xx_spi_transceive_sync(const struct device *dev, const struct spi_config *config,
+				     const struct spi_buf_set *tx_bufs,
+				     const struct spi_buf_set *rx_bufs)
+{
+	int ret = -EINVAL;
+
+	uint32_t tx_buf_count = 0;
+	uint32_t rx_buf_count = 0;
+
+	if (tx_bufs != NULL) {
+		tx_buf_count = tx_bufs->count;
+	}
+
+	if (rx_bufs != NULL) {
+		rx_buf_count = rx_bufs->count;
+	}
+
+	struct sy1xx_spi_xfer request = {
+		.cs = 0,
+		.addr_width = 0,
+		.addr = 0,
+
+		.data = 0,
+		.data_len = 0,
+	};
+
+	if ((tx_buf_count == 0) && (rx_buf_count == 1)) {
+		/* pure read */
+
+		request.data_len = rx_bufs->buffers[0].len;
+		request.data = rx_bufs->buffers[0].buf;
+
+		ret = sy1xx_spi_read(dev, &request);
+
+	} else if ((tx_buf_count == 1) && (rx_buf_count == 0)) {
+		/* pure write */
+
+		request.data_len = tx_bufs->buffers[0].len;
+		request.data = tx_bufs->buffers[0].buf;
+
+		ret = sy1xx_spi_write(dev, &request);
+
+	} else if ((tx_buf_count == 1) && (rx_buf_count == 2)) {
+		/* read from address */
+
+		/* decode address */
+		request.addr_width = 8;
+		request.addr = ((uint8_t *)tx_bufs->buffers[0].buf)[0];
+
+		/* decode data */
+		request.data = (uint8_t *)rx_bufs->buffers[1].buf;
+		request.data_len = rx_bufs->buffers[1].len;
+
+		ret = sy1xx_spi_read(dev, &request);
+
+	} else if ((tx_buf_count == 2) && (rx_buf_count == 0)) {
+		/* write to address */
+
+		/* decode address */
+		request.addr_width = 8;
+		request.addr = ((uint8_t *)tx_bufs->buffers[0].buf)[0];
+
+		/* decode data */
+		request.data = (uint8_t *)tx_bufs->buffers[1].buf;
+		request.data_len = tx_bufs->buffers[1].len;
+
+		ret = sy1xx_spi_write(dev, &request);
+	} else {
+		LOG_ERR("Unsupported spi transfer mode");
```

Comment:
No. All type of relevant transfers should be supported. TX/RX spi buffer are scatter-gather based and asymmetrical, so up to your driver to loop until all is consumed.

Reply:
Unfortunately the generic spi does not support address and data split up.